### PR TITLE
8280136: Serial: Remove unnecessary use of ExpandHeap_lock

### PR DESCRIPTION
**Pull request by albertnetymk**
This PR consists of two commits:

1. remove `ExpandHeap_lock` in Serial GC code.
2. rename it to `ParallelExpandHeap_lock` to indicate it's Parallel-GC only.

Test: tier1-6

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8280136](https://bugs.openjdk.java.net/browse/JDK-8280136): Serial: Remove unnecessary use of ExpandHeap_lock

### Reviewers
 * [Ivan Walulya](https://openjdk.java.net/census#iwalulya) (@walulyai - **Reviewer**) ⚠️ Review applies to 16874ed0256116521d9cfd26be0182b72999dcb4
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7124/head:pull/7124` \
`$ git checkout pull/7124`

Update a local copy of the PR: \
`$ git checkout pull/7124` \
`$ git pull https://git.openjdk.java.net/jdk pull/7124/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7124`

View PR using the GUI difftool: \
`$ git pr show -t 7124`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7124.diff">https://git.openjdk.java.net/jdk/pull/7124.diff</a>

</details>

### CODE REVIEW

**Review by walulyai — APPROVED** [review]
lgtm!

**Review by kimbarrett — COMMENTED** [review]
I'm not keen on the suggested new name. I want to read ParallelExpandHeap_lock
as a lock for parallel expansion of the heap, which doesn't really have the
right flavor of subsystem ownership. I think (with this change) its uses are
limited to ParallelGC oldgen expansion, suggesting a name like
PSOldGenExpand_lock (or PSOldGen::Expand_lock, which could be private were it
not for the assert in MutableSpace and the "usual" practice of defining locks
in mutexLocker.[ch]pp).

Other than that naming issue, nice cleanup.

**Comment by albertnetymk** [comment]
I have moved the mutex inside `PSOldGen` to reduce its scope.

File: src/hotspot/share/gc/parallel/psOldGen.cpp [hunk 1]
```diff
@@ -40,7 +40,12 @@
 PSOldGen::PSOldGen(ReservedSpace rs, size_t initial_size, size_t min_size,
                    size_t max_size, const char* perf_data_name, int level):
   _min_gen_size(min_size),
-  _max_gen_size(max_size)
+  _max_gen_size(max_size),
+#ifdef ASSERT
+  _Expand_lock(Heap_lock->rank()-1, "PSOldGenExpand_lock", true)
+#else
+  _Expand_lock(Mutex::safepoint, "PSOldGenExpand_lock", true)
+#endif
```
Comment (coleenp) [inline src/hotspot/share/gc/parallel/psOldGen.cpp:48]: As per our coding convention, this should be _expand_lock.

File: src/hotspot/share/gc/parallel/psOldGen.cpp [hunk 2]
```diff
@@ -163,14 +168,16 @@ bool PSOldGen::expand_for_allocate(size_t word_size) {
...
     // Avoid "expand storms" by rechecking available space after obtaining
     // the lock, because another thread may have already made sufficient
     // space available.  If insufficient space available, that will remain
     // true until we expand, since we have the lock.  Other threads may take
     // the space we need before we can allocate it, regardless of whether we
     // expand.  That's okay, we'll just try expanding again.
-    if (object_space()->needs_expand(word_size)) {
+    bool needs_expand =
+      pointer_delta(object_space()->end(), object_space()->top()) < word_size;
+    if (needs_expand) {
```
Comment (kstefanj) [inline src/hotspot/share/gc/parallel/psOldGen.cpp:180]: To me the old code reads better, but I guess it's a matter of taste. The predicate could be moved to PSOldGen to allow asserting that the lock is held.

**Review by kstefanj — APPROVED** [review]
Looks good, even if I preferred the old use of `needs_expand()`.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

**Comment by albertnetymk** [comment]
I have moved the lock back to `mutexLocker` and used the name `PSOldGenExpand_lock`.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Review by kstefanj — APPROVED** [review]

**Comment by albertnetymk** [comment]
Thanks for the review.

/integrate

**Pull request closed without merging**